src/source-af-packet.c

```diff
@@ -988,8 +998,19 @@ static inline int AFPWalkBlock(AFPThreadVars *ptv, struct tpacket_block_desc *pb
 {
     const int num_pkts = pbd->hdr.bh1.num_pkts;
     uint8_t *ppd = (uint8_t *)pbd + pbd->hdr.bh1.offset_to_first_pkt;
+    const int loopback = ptv->loopback;
 
     for (int i = 0; i < num_pkts; ++i) {
+        /* Drop outgoing packets on loopback interfaces, only process incoming ones. */
+        if ( loopback ) {
```

Comment:
I'm not a fan of this check for a rare edge case on a per packet basis. I don't currently have an alternative in mind though.

Reply:
Yep. While I'd be surprised the overhead is measurable, it does stand out quite prominently.

A second `AFPWalkBlockLoopback` and  function pointer indirection  `ptv->walk_block(...)` would maybe do.

But, not sure that's worth it. I don't personally have a use-case for `lo` sniffing with Suricata (cursory google indicates there are cases of people wanting to do it, e.g. reverse proxy monitoring). In Zeek duplicated packets are flagged as retransmissions and very visible in the conn.log. Also, there was/is a discrepancy how the default libpcap based packet source behaves compared to the AF_PACKET one (and IMO they should behave the same in Zeek sniffing loopback).

Given this has not come up before (or not visible/noticable enough), maybe this is just not needed. How would you feel about a warning during source initialization with something like: "Monitoring loopback interfaces with AF_PACKET results in duplicated packets, consider using the libpcap based source instead or explain why you need to use AF_PACKET :-)" instead?

Reply:
How about replacing this with something like
```c
for (int i = 0; i < num_pkts; ++i) {
   const struct sockaddr_ll *sll =  (const struct sockaddr_ll*)(ppd + TPACKET_ALIGN(sizeof(struct tpacket3_hdr)));
   if (unlikely(ptv->pkt_flags & sll->sll_pkttype) == 0)) {
       ppd = ppd + ((struct tpacket3_hdr *)ppd)->tp_next_offset;
       continue;
   }
```
It still has the check, but now it's more generalized. The Setup code would need to properly setup `ptv->pkt_flags`

Reply:
> It still has the check, but now it's more generalized. The Setup code would need to properly setup ptv->pkt_flags

That looks special, yes The values of `sll_pkttype` aren't suitable to be used with a bit mask, so that would need to look a bit more like the following with a few asserts about the max value of pkttype :-/
```
(ptv->pkt_flags & (1 << sll->sll_pkttype))
```

I can give that a whirl if that seems still okay.

Reply:
Could use `ptv->pkt_flags & BIT_U32(sll->ssl_pkttype)`. But yeah lets give this a try.